### PR TITLE
Async Indicator metrics

### PR DESCRIPTION
**Pull request by emord**
Does a bit of a refactor and adds some metrics for async indicators

* metric for number of docs in the queue tagged with the config ids
* histogram for amount of time processing a document takes tagged with the config ids

buddy @esoergel

### CODE REVIEW

File: corehq/apps/userreports/migrations/0007_index_indicator_config_ids.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    dependencies = [
+        ('userreports', '0006_add_index_to_domain'),
+    ]
+
+    operations = [
+        HqRunSQL(
+            add_if_not_exists_raw(
+                """
+                CREATE INDEX userreports_asyncindicator_indicator_config_ids_gin_idx
```
Comment (emord) [inline corehq/apps/userreports/migrations/0007_index_indicator_config_ids.py]: Will need to add this concurrently on NIC

**Review by esoergel — APPROVED** [review]
That task query is wicked confusing, but I'm convinced.

**Comment by esoergel** [comment]
```
======================================================================
FAIL: corehq.apps.userreports.tests.test_pillow:AsyncIndicatorTest.test_async_save_fails
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/vendor/lib/python2.7/site-packages/mock.py", line 1224, in patched
    return func(*args, **keywargs)
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/tests/test_pillow.py", line 463, in test_async_save_fails
    queue_async_indicators()
  File "/vendor/lib/python2.7/site-packages/celery/app/task.py", line 420, in __call__
    return self.run(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/tasks.py", line 226, in queue_async_indicators
    _async_indicator_metric('commcare.async_indicator.oldest_created_indicator', lag)
  File "/mnt/commcare-hq-ro/corehq/apps/userreports/tasks.py", line 45, in _async_indicator_metric
    datadog_gauge(name, value, enforce_prefix='commcare.async_indicator', tags=tags)
  File "/mnt/commcare-hq-ro/corehq/util/datadog/gauges.py", line 37, in datadog_gauge
    _datadog_record(statsd.gauge, name, value, enforce_prefix, tags)
  File "/mnt/commcare-hq-ro/corehq/util/datadog/gauges.py", line 50, in _datadog_record
    _enforce_prefix(name, enforce_prefix)
  File "/mnt/commcare-hq-ro/corehq/util/datadog/gauges.py", line 79, in _enforce_prefix
    "pass enforce_prefix=None".format(name))
  File "/mnt/commcare-hq-ro/corehq/util/soft_assert/core.py", line 34, in __call__
    return self._call(assertion, msg, obj)
  File "/mnt/commcare-hq-ro/corehq/util/soft_assert/core.py", line 48, in _call
    raise AssertionError(msg)
AssertionError: Did you mean to call your gauge 'commcare.commcare.async_indicator.oldest_created_indicator'? If you're sure you want to forgo the prefix, you can pass enforce_prefix=None
```

**Comment by emord** [comment]
@esoergel hopefully https://github.com/dimagi/commcare-hq/pull/16961/commits/835ad05b403f2a39ca55e33d0b47d1e5b3ebfd2c makes it a little better.

Planning to add more metrics in later PRs

File: corehq/apps/userreports/tasks.py [hunk 2]
```diff
@@ -277,43 +280,84 @@ def save_document(doc_ids):
...
+        datadog_gauge('commcare.async_indicator.indicator_count', count, tags=config_ids)
+
+
+def _indicators_by_count():
+    """
+    Number of docs in the queue that have a specific set of indicator config ids
+
+    query returns
+    [
+      {"indicator_config_ids": [data_source_id_1, data_source_id_1], indicator_config_ids__count: "12"}
```
Comment (emord) [inline corehq/apps/userreports/tasks.py]: Note for me tomorrow. Make this return {datasource_id: count} as indicator_config_ids can be out of order

Comment (esoergel) [inline corehq/apps/userreports/tasks.py]: Nice, thanks for the docs!

File: corehq/apps/userreports/tasks.py [hunk 3]
```diff
@@ -365,3 +365,16 @@ def _indicators_by_count(date_created=None):
...
 
     return ret
+
+
+@periodic_task(
+    run_every=crontab(minute="*/15"),
+    queue=settings.CELERY_PERIODIC_QUEUE,
+)
+def icds_async_indicators_metrics():
+    indicator_count_until_28 = _indicators_by_count(datetime(2017, 6, 28))
```
Comment (esoergel) [inline corehq/apps/userreports/tasks.py]: what's with the specific date?

**Review by esoergel — APPROVED** [review]

File: corehq/apps/userreports/tasks.py (hunk 3 above)
Comment (emord) [inline corehq/apps/userreports/tasks.py]: i'm going to use it for a dashboard for a rebuild happening on icds. the 28th is a super safe date to use. meaning that i'm 100% ceratin that everything that matters in the table is before that date. I think I could use a later date, but I don't have anything that can verify what date is the safest..

**Comment by emord** [comment]
@esoergel added a couple of commits. I think this is good to merge

**Review by esoergel — APPROVED** [review]

**Pull request merged**